pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,835 @@
+
+        >>> new_tape = tape.expand()
+        >>> new_tape.operations
+        [PauliX(wires=[0]),
+         PauliX(wires=['a']),
+         Rot(0.543, 0.1, 0.4, wires=[0]),
+         CNOT(wires=[0, 'a']),
+         RY(0.2, wires=['a'])]
+        """
+        new_tape = expand_tape(
+            self, depth=depth, stop_at=stop_at, expand_measurements=expand_measurements
+        )
+        new_tape._update()
+        return new_tape
+
+    def inv(self):
+        """Inverts the processed operations.
+
+        Inversion is performed in-place.
+
+        .. note::
+
+            This method only inverts the quantum operations/unitary recorded
+            by the quantum tape; state preparations and measurements are left unchanged.
+
+        **Example**
+
+        .. code-block:: python
+
+            with QuantumTape() as tape:
+                qml.BasisState(np.array([1, 1]), wires=[0, 'a'])
+                qml.RX(0.432, wires=0)
+                qml.Rot(0.543, 0.1, 0.4, wires=0).inv()
+                qml.CNOT(wires=[0, 'a'])
+                probs(wires=0), probs(wires='a')
+
+        This tape has the following properties:
+
+        >>> tape.operations
+        [BasisState(array([1, 1]), wires=[0, 'a']),
+         RX(0.432, wires=[0]),
+         Rot.inv(0.543, 0.1, 0.4, wires=[0]),
+         CNOT(wires=[0, 'a'])]
+        >>> tape.get_parameters()
+        [array([1, 1]), 0.432, 0.543, 0.1, 0.4]
+
+        Here, let's set some trainable parameters:
+
+        >>> tape.trainable_params = {1, 2}
+        >>> tape.get_parameters()
+        [0.432, 0.543]
+
+        Inverting the tape:
+
+        >>> tape.inv()
+        >>> tape.operations
+        [BasisState(array([1, 1]), wires=[0, 'a']),
+         CNOT.inv(wires=[0, 'a']),
+         Rot(0.543, 0.1, 0.4, wires=[0]),
+         RX.inv(0.432, wires=[0])]
+
+        Tape inversion also modifies the order of tape parameters:
+
+        >>> tape.get_parameters(trainable_only=False)
+        [array([1, 1]), 0.543, 0.1, 0.4, 0.432]
+        >>> tape.get_parameters(trainable_only=True)
+        [0.543, 0.432]
+        >>> tape.trainable_params
+        {1, 4}
+        """
+        # we must remap the old parameter
+        # indices to the new ones after the operation order is reversed.
+        parameter_indices = []
+        param_count = 0
+
+        for queue in [self._prep, self._ops, self.observables]:
+            # iterate through all queues
+
+            obj_params = []
+
+            for obj in queue:
+                # index the number of parameters on each operation
+                num_obj_params = len(obj.data)
+                obj_params.append(list(range(param_count, param_count + num_obj_params)))
+
+                # keep track of the total number of parameters encountered so far
+                param_count += num_obj_params
+
+            if queue == self._ops:
+                # reverse the list representing operator parameters
+                obj_params = obj_params[::-1]
```

Comment:
Later one could change the parameters whose inverse gate just take ``(-1) param`` here...

Reply:
Oh nice idea!

We would have to have a way for operations to declare _how_ they are invertible; it could be either a

* linear combination of parameters (includes the case of `-param[0], param[1], param[2]`)

* something more complicated (e.g., `param[0].conj().T` for `QubitUnitary`)

* no parameters; cannot be inverted this way, relies on the device implementation.